### PR TITLE
[#10470] Add v6 instructor join link support

### LINKED ISSUE #10470: V6 instructor join link incompatible with V7
Problems:
1. After switching to V7, instructor join links sent by V6 will no longer work. This is going to affect those who had accounts created just prior to the switchover.
2. Even those who have joined already, sometimes dig up the old join link and click on it simply as a way to reach TEAMMATES. Will that continue to work?

Is it possible to bake in something into V7 to take care of the above, at least fail gracefully and guide user to v7?

### PR DESCRIPTION
**Pull request by madanalogy**
Fixes #10470 

**Outline of Solution**: Add institute mac when institute is present during routing from old join link to new one

### CODE REVIEW

**Comment by damithc** [comment]
`invalid` -> `invalid or expired`? When we are near to the switchover, I'll inform new users to use the link within 24 hours or it will expire. 24 hours prior to the switchover, I'll pause creating new accounts until we switch over.

If the user has joined already and clicks on the V6 join link again, goes to the home page (i.e., this error message not shown), right?

**Comment by wkurniawan07** [comment]
This is unnecessary. The routing is already there.

**Comment by madanalogy** [comment]
> This is unnecessary. The routing is already there.

I'll just remove my second commit then?

**Comment by madanalogy** [comment]
> `invalid` -> `invalid or expired`? 

There is actually no way to tell if I'm not wrong.

> When we are near to the switchover, I'll inform new users to use the link within 24 hours or it will expire. 24 hours prior to the switchover, I'll pause creating new accounts until we switch over.

Sounds like a good measure.

> If the user has joined already and clicks on the V6 join link again, goes to the home page (i.e., this error message not shown), right?

I'm actually not sure on this, I don't think so? But then again, I don't think it does for v6 either? Maybe @wkurniawan07 can clarify

File: src/web/app/user-join-page.component.ts [hunk 1]
```diff
@@ -18,6 +18,8 @@ import { ErrorMessageOutput } from './error-message-output';
 })
 export class UserJoinPageComponent implements OnInit {
 
+  link: string = '/web/front/request';
```
Comment (xpdavid) [inline src/web/app/user-join-page.component.ts:21]: You can just put this into the the template html.

File: src/web/app/user-join-page.component.html [hunk 2]
```diff
@@ -12,7 +12,8 @@ <h3>Not logged in</h3>
       <h3>Invalid course join link</h3>
     </div>
     <div class="card-body" *ngIf="!validUrl">
-      You are using an invalid join link. Please <a [routerLink]="link">request for a new join link</a> or check that you copied the URL entirely.
+      You are using an invalid join link. Please <a routerLink="/web/front/request">request for a new join link</a>
+      or check that you copied the URL entirely.
```
Comment (wkurniawan07) [inline src/web/app/user-join-page.component.html:16]: Requesting a new join link is not the same as requesting a new account; the latter creates a duplicate sample course.

File: src/web/app/app.module.ts [hunk 3]
```diff
@@ -84,6 +84,12 @@ const routes: Routes = [
       },
     ],
   },
+  {
+    // This path is added to support redirection of v6 join link to v7
+    path: 'page/instructorCourseJoin',
+    pathMatch: 'full',
+    redirectTo: 'web/join',
+  },
```
Comment (wkurniawan07) [inline src/web/app/app.module.ts:92]: Not necessary; this can be removed.

**Comment by damithc** [comment]
> Given the short time period we have, the first solution should be the one to go.

I think so too. We didn't have institution validation all this time either. No harm bypassing it for these special cases.

File: src/web/app/user-join-page.component.html (hunk 2 above)
Comment (madanalogy) [reply src/web/app/user-join-page.component.html:16]: So should I remove the redirection to the request page then? I'll create an issue separately for regenerating join links

**Comment by damithc** [comment]
@madanalogy if you are taking the _bypassing temporarily_ route, you may want to add a comment in the code to inform future readers of the code.

File: src/main/java/teammates/ui/webapi/action/LegacyUrlMapper.java [hunk 4]
```diff
@@ -31,10 +33,17 @@ public void doGet(HttpServletRequest req, HttpServletResponse resp) throws IOExc
...
         case Const.LegacyURIs.INSTRUCTOR_COURSE_JOIN:
             key = req.getParameter(Const.ParamsNames.REGKEY);
-            redirectUrl = Config.getFrontEndAppUrl(Const.WebPageURIs.JOIN_PAGE)
+            String institute = req.getParameter(Const.ParamsNames.INSTRUCTOR_INSTITUTION);
+            AppUrl newUrl = Config.getFrontEndAppUrl(Const.WebPageURIs.JOIN_PAGE)
                     .withRegistrationKey(key)
-                    .withParam(Const.ParamsNames.ENTITY_TYPE, Const.EntityType.INSTRUCTOR)
-                    .toString();
+                    .withParam(Const.ParamsNames.ENTITY_TYPE, Const.EntityType.INSTRUCTOR);
+            // TODO: Remove the following code block for mac generation once v6 join links stop being supported
```
Comment (wkurniawan07) [inline src/main/java/teammates/ui/webapi/action/LegacyUrlMapper.java:40]: Not necessary. This entire class will be removed in the first place one V6 is decommissioned.

**Pull request merged**